usage.go

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) 2016 Kelsey Hightower and others. All rights reserved.
+// Use of this source code is governed by the MIT License that can be found in
+// the LICENSE file.
+
+package envconfig
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"reflect"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+	"text/template"
+)
+
+const (
+	// NoHeader constant to use for no header on usage output
+	NoHeader = ""
+
+	// DefaultHeader constant to use for default header on usage output
+	DefaultHeader = `USAGE: {{ . }}
+  This application is configured via the environment. The following environment
+  variables can used specified:
+
+`
+	// DefaultListFormat constant to use to display usage in a list format
+	DefaultListFormat = `  {{.Key}}
+    [description] {{.Description}}
+    [type]        {{.Type}}
+    [default]     {{.Default}}
+    [required]    {{.Required}}`
+
+	// DefaultTableFormat constant to use to display usage in a tabluar format
+	DefaultTableFormat = "table  {{.Key}}\t{{.Type}}\t{{.Default}}\t{{.Required}}\t{{.Description}}"
+)
+
+// usageInfo used to provide values for tabular output to the template function
+type usageInfo struct {
+	Key         string
+	Description string
+	Type        string
+	Default     string
+	Required    string
+}
+
+// toTypeDescription convert techie type information into something more human
+func toTypeDescription(t reflect.Type) string {
+	switch t.Kind() {
+	case reflect.Array, reflect.Slice:
+		return fmt.Sprintf("List of %s", toTypeDescription(t.Elem()))
+	case reflect.Ptr:
+		return fmt.Sprintf("Reference to %s", toTypeDescription(t.Elem()))
+	case reflect.Chan:
+		return fmt.Sprintf("Channel of %s", toTypeDescription(t.Elem()))
+	case reflect.Map:
+		return fmt.Sprintf("Map of %s to %s", toTypeDescription(t.Key()), toTypeDescription(t.Elem()))
+	case reflect.Func:
+		return "Function"
+	case reflect.Interface:
+		return "Interface"
+	case reflect.Struct:
+		if t.Name() != "" {
+			return t.Name()
+		}
+		return "Nested Struct"
+	case reflect.String:
+		name := t.Name()
+		if name != "" && name != "string" {
+			return name
+		}
+		return "String"
+	case reflect.Bool:
+		name := t.Name()
+		if name != "" && name != "bool" {
+			return name
+		}
+		return "Boolean"
+	case reflect.Int:
+		name := t.Name()
+		if name != "" && name != "int" {
+			return name
+		}
+		return "Integer"
+	case reflect.Int8:
+		name := t.Name()
+		if name != "" && name != "int8" {
+			return name
+		}
+		return "Integer(8 bits)"
+	case reflect.Int16:
+		name := t.Name()
+		if name != "" && name != "int16" {
+			return name
+		}
+		return "Integer(16 bits)"
+	case reflect.Int32:
+		name := t.Name()
+		if name != "" && name != "int32" {
+			return name
+		}
+		return "Integer(32 bits}"
+	case reflect.Int64:
+		name := t.Name()
+		if name != "" && name != "int64" {
+			return name
+		}
+		return "Integer(64 bits)"
```

Comment:
I will collapse them. If people complain they want the bit specificity, we can always add it back.